icefoganalytics/travel-authorization
# Add Support for Estimates Generation on Estimate Tab

References
- [Modeling, UI Label To Data Model Mapping for Travel Authorization Form, 2023-10-05](https://docs.google.com/document/d/1SXCUgpUMgFLESLaHZa-Duv03zzJbRDZ2BJCxmN7GXao/edit?pli=1)
- [Modeling, Trip Segment to Accommodation Estimate, 2023-10-12](https://docs.google.com/document/d/13CmY-0ijYIgRfvELSppSQmYlGM52APaG9ZagMfY7DSQ/edit)

Fixes https://github.com/icefoganalytics/travel-authorization/issues/18

# User Story

As a User,
I want the system to auto-create estimates for my trip from my draft travel authorization request data,
So that I can speed up my travel authorization request writing.

# Context

On the `/my-travel-request/:form-id/` page, there is a tab for "Estimates", before the user has entered any estimates this page should have a button to "generate estimates". This button should replace the "Add Estimate" button, when the table has no content.

![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/f76ea01a-b6a7-42c9-b7e3-ded8831d490f)

## Examples

Estimates
Flights:
1 stop away (Vancouver, Kelowna, Edmonton, Calgary, Yellowknife): $350 per segment
Not above destinations: $1000 per segment

Private Vehicle:
Use distance between Yukon places (distance api) x private vehicle km rate (api)

Hotels: 
All hotels default to $250/night

Private Accommodations:
Use perdiem apis

Per diems:
Based on perdiem api already created.
If only dates are entered calculate based on full day
if times entered calculate based on times entered if should include

# Screenshots

Pre-Estimate Generation
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/fac74364-234e-4a4d-b414-704dc5c9e2fe)

Round Trip - after estimate generation
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/e1e5c862-0785-4a0a-a583-d66d83050482)
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/d2d6d5cf-d640-4fc5-b409-17613e8fa7fc)

One Way - after estimate generation with missing distance data
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/ea27fe16-ccdc-4de8-a8cb-8e742f00b4e9)
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/c6dc1229-ca78-4bc2-aabb-f3f0882834af)

One Way - after estimate generate with good data
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/183dfb1c-6f45-49be-8e3b-84fd6be053e7)
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/3d5e84c2-c028-4d1e-b725-12ea128c19e8)

Multi-Destination - after estimate generation
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/c07cea7c-90f5-466c-849d-96dd80bdc65a)
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/af90ef2b-4ab0-44fa-83bd-49296fc0d8ea)

# Testing Instructions
1. Boot the database service via `dev up`
2. Boot the api service via `npm run start`
3. Boot the web service via `npm run start`
4. Log in to the app at http://localhost:8080/
5. Go to the "My Travel Requests" page via the top nav (http://localhost:8080/my-travel-requests)
6. Create a new travel request via the "+ Travel Authorization" button.
7. Set up the "Details" section of the "Details" tab as a "Round Trip".
8. Go from "Whitehorse" to "Vancouver" on the 5 of any month.
9. Travel by "Aircraft" and stay in a "Hotel.
10. Go from "Vancouver" to "Whitehorse" via "Aircraft".
11. Leave the return accommodation blank, or clear it as needed.
12. Save the form via the "Save Draft" button at the bottom.
13. Go to the "Estimates" tab of the draft.
14. You will now see a "Generate Estimates" button, click it and confirm that you want to generate some estimates.
15. Check that the estimates look like you'd expect them too, they should be vaguely similar to the wireframe, with some minor changes due to data constraints. If anything looks really weird though is probably a bug.
16. You will get different Meals & Incidentals results depending on what times you selected as leaving/returning at.
17. Check that the "Generate Estimates" button has been replaced by the "Add Estimate" button.
18. Delete all estimates that currently exist, and the "Generate Estimates" button will return.
19. Go back to the "My Travel Requests" page and create a new Travel Request.
20. Set your trip type as One Way.
21. Go from Whitehorse to Red Deer (AB) on any day of the month, travel by "Personal Vehicle" and stay in "Private" accommodation.
22. Save your draft via the "Save Draft" button.
23.  Go to the "Estimate" tab.
24. Click the "Generate Estimates" button, and confirm the action.
25. Check that this generates a single row with the correct transport path, but 0 for the cost. This is because there isn't any distance data between Whitehorse and Red Deer (AB).
26. Nothing is generated for the accommodation either, as an end date for the stays is not specified. We may need to change the data model, and UI to handle this kind of thing.
27. Go back to the "Details" tab and the "Details" panel.
28. Check the "In Territory" flag.
29. Select the start location as "Carcross" and the end location as "Whitehorse".
    **Note that most places in the distance matrix don't match the location selectors** - this is the result of a lack of data normalization, and a lack of shared entities connected by foreign keys
30. Save as "Draft" again.
31. Go back to the Estimates tab, and delete the previous estimate.
32. You should again see the "Generate Estimates" button.
33. Generate a new estimate.
34. Check that the new estimate has an amount take from the `distanceMatrix` table `kilometers` field multiplied by the "distance allowance per killometer" of `0.605`
35. Go back to the "My Travel Estimates" page, and generate a new draft travel request.
36. Make a "Multi-Destination Trip"
37. Fly Whitehorse to Vancouver and stay in a hotel on the 5th of the month.
38. Go from Vancouver to Airdrie (AB) by Bus and stay in "Private" accommodations on the 7th.
39. Go from Airdrie (AB) to Loyalist Township (ON) by Aircraft on the 17th and say in a Hotel.
40. Save your draft.
41. Go to the Estimates tab and generate some estimates.
42. There will be a lot of them (3 pages) on my count. Note that the total count is still valid because its loading all content into the front-end, and doing the computation there as well. Future work should push total count computation to the back-end, and only load each page as needed.
43. Check that these seem to be more or less correct.
44. Create a new draft travel authorization and then ...
45. Pick a bunch of random stuff and see if it breaks. It probably will. Without any tests, its virtually impossible to validate a generation system of this complexity.
